torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,272 @@
+//
+// How does the compiler decide which one to choose? According to `https://en.cppreference.com/w/cpp/language/list_initialization`,
+// braced-init-list always prefers the constructor that takes `std::initializer_list`. Hence we happily
+// move forward with constructor #2, and it calls the following:
+//
+// `TensorDataContainer(1)`
+//
+// Now it matches `TensorDataContainer(int value)`, which stores `1` as a scalar value. All is good.
+//
+// Note that `torch::tensor({{1}, {2}})` can also match another previously existing function overload:
+// `torch::tensor(at::ArrayRef<int> values)`, because `{1}` and `{2}` can be treated as
+// a list-initialization of an `int` value. However, this will produce a Tensor with sizes `{2}`,
+// but we actually want a Tensor with sizes `{2, 1}`. In order to avoid matching this function overload,
+// we removed the function overload and moved the ability to convert `at::ArrayRef<T>` (and similarly `std::vector<T>`)
+// into `TensorDataContainer`, and since for braced-init-list the `TensorDataContainer(std::initializer_list<TensorDataContainer>)`
+// constructor is always preferred over all other constructors, it will take the `std::initializer_list` path
+// and all is good again.
+struct TensorDataContainer {
+  // NOTE: For tensors with zero-size dimensions (e.g. `torch::tensor({{}, {}})`),
+  // the innermost empty braced-init-list `{}` matches the default constructor of
+  // the innermost `TensorDataContainer`.
+  TensorDataContainer() : sizes_({0}), scalar_type_(at::ScalarType::Undefined), type_(TensorDataContainerType::InitList) {}
+#define TENSOR(T, S) \
+  TensorDataContainer(T value) : sizes_(), scalar_type_(at::k##S), type_(TensorDataContainerType::Scalar), scalar_(value) {}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+  TensorDataContainer(std::initializer_list<TensorDataContainer> init_list) :
+      sizes_(),
+      scalar_type_(init_list.begin()->scalar_type()),
+      type_(TensorDataContainerType::InitList),
+      init_list_(init_list) {
+    const TensorDataContainer& first_elem = *(init_list.begin());
+    for (const auto& elem : init_list) {
+      TORCH_CHECK(elem.sizes() == first_elem.sizes(),
+        "Expected all sub-lists to have sizes: ",
+        first_elem.sizes(),
+        " (e.g. ", first_elem, "), ",
+        "but got sub-list ",
+        elem,
+        " with sizes: ",
+        elem.sizes());
+      TORCH_CHECK(elem.scalar_type() == first_elem.scalar_type(),
+        "Expected all elements of the tensor to have the same scalar type: ",
+        first_elem.scalar_type(),
+        ", but got element of scalar type: ",
+        elem.scalar_type());
+    }
+    sizes_.clear();
+    sizes_.reserve(first_elem.sizes().size() + 1);
+    sizes_.push_back(init_list.size());
+    sizes_.insert(sizes_.end(), first_elem.sizes().begin(), first_elem.sizes().end());
+  }
+
+#define TENSOR(T, S) \
+  TensorDataContainer(const at::ArrayRef<T>& values) { \
```

Comment:
I think ArrayRef is meant to be passed by value

Reply:
Fixed